File: filters/split_into_topics.lua
local current_section
local sections = {}
local target_directory


function create_target_directory()
  local filepath = PANDOC_STATE['output_file'] 
  target_directory, __ = pandoc.path.split_extension(filepath) 
  pandoc.system.make_directory (target_directory, true)
end

function Header(elem)
    if elem.level == 1 then
        if current_section then
            table.insert(sections, current_section)
        end
        current_section = {title = pandoc.utils.stringify(elem), blocks = {}}
    end
    return elem
end

function Block(elem)
    if current_section then
        table.insert(current_section.blocks, elem)
    end
    return elem
end

function Pandoc(doc)
    create_target_directory() 
    if current_section then
        table.insert(sections, current_section)
    end
    for i, section in ipairs(sections) do 
        local new_doc = pandoc.Pandoc(section.blocks, doc.meta)
        local filepath = pandoc.path.join({target_directory, section.title})
        pandoc.pipe("pandoc", {"-o", filepath..".md"}, pandoc.write(new_doc, "markdown"))
    end
    os.exit()
end

File: filters/split_into_topics.py
#!/home/jeremy/Python3.10Env/bin/python
# -*- coding: utf-8 -*-
#
#  module.py
#
#  Copyright 2023 Jeremy Allan <jeremy@jeremyallan.com>
import panflute as pf
from pathlib import Path



def set_vars(doc):
  metadata = doc.get_metadata()
  doc.source = Path(metadata['inputfile'])
  doc.parent = Path(metadata['outputfile'])
  doc.sections = []


def load_sections(elem, doc):
  if isinstance(elem, pf.Header) and elem.level == 1:
    doc.sections.append(dict(header=elem, content=[])) 
  else:
    try:
      doc.sections[-1]['content'].append(elem)
    except:
      pass
  return elem

def write_sections(doc):
  """Writes each section content to a separate markdown file.

  This function assumes `doc.sections` is populated with section content lists.
  """
  for section in doc.sections:
    doc.parent.mkdir(parents=True, exist_ok=True)
    filepath = doc.parent.joinpath(pf.stringify(section['header'])).with_suffix('.md')
    args = [f'--output={filepath}']
    pf.run_pandoc('This is a test', args=args)
    # with filepath.open('w') as fp:
    #   fp.writelines(pf.stringify(section['content']))

def main(doc=None):
  return pf.run_filter(load_sections, prepare=set_vars, finalize=write_sections, doc=doc) 
 

if __name__ == '__main__':
  main()
 